src/main/java/bio/terra/service/filedata/VirtualFileSystemUtils.java

```diff
@@ -0,0 +1,157 @@
+package bio.terra.service.filedata;
+
+import bio.terra.service.filedata.exception.DirectoryMetadataComputeException;
+import bio.terra.service.filedata.exception.FileNotFoundException;
+import bio.terra.service.filedata.google.firestore.FireStoreDirectoryEntry;
+import bio.terra.service.filedata.google.firestore.FireStoreFile;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.apache.commons.codec.digest.DigestUtils;
+import org.apache.commons.codec.digest.PureJavaCrc32C;
+import org.apache.commons.collections4.CollectionUtils;
+import org.apache.commons.lang3.StringUtils;
+
+public class VirtualFileSystemUtils {
```

Comment:
Does it make sense to put these in `FileMetadataUtils.java`? That is meant to be the shared code that was originally in FirestoreUtils

Reply:
It might. I'll poke around

Reply:
I think it belongs in a separate class, since its very specific logic. I agree that the name isn't great. I'll change it.